cpp/src/join/conditional_join.cu

```diff
@@ -48,9 +48,7 @@ std::unique_ptr<rmm::device_uvector<size_type>> conditional_join_anti_semi(
 {
   if (right.num_rows() == 0) {
     switch (join_type) {
-      case join_kind::LEFT_ANTI_JOIN:
-        return get_trivial_left_join_indices(left, stream, rmm::mr::get_current_device_resource())
-          .first;
+      case join_kind::LEFT_ANTI_JOIN: return get_trivial_left_join_indices(left, stream, mr).first;
```

Comment:
Humnn, this seems wrong to me, since we don't return `rvalue` thus `*.first` seems to be returned by copied. However, `*.first` here actually is a `unique_ptr` thus it can't be copied. I'm surprised that this code can compile.

Reply:
I'm not sure if I follow. I copied this from here: https://github.com/rapidsai/cudf/blob/cdfb550f442e846623c721082128a095f02efff9/cpp/src/join/mixed_join_semi.cu#L120-L122

Reply:
I meant the C++ syntax looks like it is wrong but it isn't (because it can compile) so I'm confused...

Reply:
The return may be ellided https://en.cppreference.com/w/cpp/language/copy_elision -- return value optimization (RVO) by the compiler. If the compiler cannot do this it will report an error since you cannot copy a unique-ptr. So no error means RVO was achieved.

Reply:
It seems that the compiler automatically optimizes the case of `unique_ptr` for us. I did some experiments here: https://godbolt.org/z/zs3TsW1do

So if we don't use `unique_ptr`, the returned object is copied. When put in `unique_ptr` wrapper, there is no copy at all.

Reply:
This came up in another PR recently. As David mentioned, the RVO is technically optional for the compiler, but under the circumstances the only other option is a compiler error because the type isn't copyable, so it's not really a problem to rely on this behavior because we'll immediately know if it doesn't do what we want if the code doesn't compile.

Reply:
This behavior is good in general. However, correct me if I'm wrong, the situations under which RVO is performed are not fully defined by C++ standards. In other words, this can compile this time but may fail to compile in the future version of GCC if that version changes its RVO behavior.

Reply:
You are correct that it is optional. The worst case outcome is an easy-to-diagnose failure to compile rather than a successful compilation with incorrect (or undefined) behavior, so I don't think any action is strictly necessary. I'd certainly be open to being more explicit, but as David pointed out in https://github.com/rapidsai/cudf/pull/15645#discussion_r1610770021 we use this pattern in many places in libcudf (as evidenced by the fact that Bradley copy-pasted it), so if we were to make any changes we would want to do so systematically.